db/cache_db.go

```diff
@@ -0,0 +1,157 @@
+package db
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+)
+
+// If value is nil but deleted is false,
+// it means the parent doesn't have the key.
+// (No need to delete upon Commit())
+type cDBValue struct {
+	value   []byte
+	deleted bool
+}
+
+// CacheDB wraps an in-memory cache around an underlying DB.
+type CacheDB struct {
+	mtx    sync.Mutex
+	cache  map[string]cDBValue
+	parent DB
+}
+
+// NOTE: Needed by MultiStore.CacheWrap().
+var _ atomicSetDeleter = (*CacheDB)(nil)
+
+func NewCacheDB(parent DB) *CacheDB {
+	db := &CacheDB{
+		cache:  make(map[string]cDBValue),
+		parent: parent,
+	}
+	return db
+}
+
+func (db *CacheDB) Get(key []byte) []byte {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	value, ok := db.cache[string(key)]
+	if !ok {
+		data := db.parent.Get(key)
+		value = cDBValue{value: data}
+		db.cache[string(key)] = value
+	}
+	return value.value
+}
+
+func (db *CacheDB) Set(key []byte, value []byte) {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	db.SetNoLock(key, value)
+}
+
+func (db *CacheDB) SetSync(key []byte, value []byte) {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	db.SetNoLock(key, value)
+}
+
+func (db *CacheDB) SetNoLock(key []byte, value []byte) {
+	db.cache[string(key)] = cDBValue{value, false}
+}
+
+func (db *CacheDB) Delete(key []byte) {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	db.DeleteNoLock(key)
+}
+
+func (db *CacheDB) DeleteSync(key []byte) {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	db.DeleteNoLock(key)
+}
+
+func (db *CacheDB) DeleteNoLock(key []byte) {
+	db.cache[string(key)] = cDBValue{nil, true}
+}
+
+func (db *CacheDB) Close() {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	db.parent.Close()
+}
+
+func (db *CacheDB) Print() {
```

Comment:
So while Print is useful from an external perspective, it is not that easy to test. IMHO we could do something like
```go
func (db *CacheDB) Fprint(w io.Writer) {
   db.mtx.Lock()
   defer db.mtx.Unlock()

   fmt.Fprintln(w, "CacheDB\ncache:")
   for key, value := range db.cache {
         fmt.Fprintf(w, "[%X]:\t[%v\n", key, value)
   }
   fmt.Fprintln(w, "\nparent:")
   db.parent.Print()
}

func (db *CacheDB) Print() { db.Fprint(os.Stdout) }
```

and then properly test the order and expected out.

Reply:
I wouldn't want to test Print functions TBH.